include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -895,6 +1021,12 @@ namespace VectorTools
                 const Quadrature<dim-1>        &q_boundary,
                 const bool                     project_to_boundary_first)
   {
+    Assert((dynamic_cast<const parallel::shared::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcMessage("Use project_distributed() for parallel Triangulations!"));
+    Assert((dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcMessage("Use project_distributed() for parallel Triangulations!"));
```

Comment:
You can unify these two asserts if you do the `dynamic_cast` to `parallel::Triangulation`. That also takes care of possible future triangulation classes that happen to be parallel.